File: nim_core_v2/android/src/main/kotlin/com/netease/nimflutter/services/FLTChatroomMessageCreatorService.kt
/*
 * Copyright (c) 2022 NetEase, Inc. All rights reserved.
 * Use of this source code is governed by a MIT license that can be
 * found in the LICENSE file.
 */

package com.netease.nimflutter.services

import android.content.Context
import com.netease.nimflutter.FLTService
import com.netease.nimflutter.LocalError.createMessageFailed
import com.netease.nimflutter.NimCore
import com.netease.nimflutter.NimResult
import com.netease.nimflutter.extension.toMap
import com.netease.nimflutter.extension.toV2NIMChatroomMessage
import com.netease.nimflutter.toMap
import com.netease.nimlib.sdk.v2.chatroom.V2NIMChatroomMessageCreator
import com.netease.nimlib.sdk.v2.chatroom.model.V2NIMChatroomMessage
import com.netease.nimlib.sdk.v2.message.V2NIMMessageAttachmentCreator
import com.netease.nimlib.sdk.v2.message.attachment.V2NIMMessageCustomAttachment
import kotlin.coroutines.resume
import kotlinx.coroutines.suspendCancellableCoroutine

class FLTChatroomMessageCreatorService(
    applicationContext: Context,
    nimCore: NimCore
) : FLTService(applicationContext, nimCore) {
    override val serviceName: String = "V2NIMChatroomMessageCreator"

    init {
        nimCore.onInitialized {
            registerFlutterMethodCalls(
                "createTextMessage" to this::createTextMessage,
                "createImageMessage" to this::createImageMessage,
                "createAudioMessage" to this::createAudioMessage,
                "createVideoMessage" to this::createVideoMessage,
                "createFileMessage" to this::createFileMessage,
                "createLocationMessage" to this::createLocationMessage,
                "createCustomMessage" to this::createCustomMessage,
                "createTipsMessage" to this::createTipsMessage,
                "createCustomMessageWithAttachment" to this::createCustomMessageWithAttachment,
                "createCustomMessageWithAttachmentAndSubType" to this::createCustomMessageWithAttachmentAndSubType,
                "createForwardMessage" to this::createForwardMessage
            )
        }
    }

    private suspend fun createTextMessage(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            val text = arguments["text"] as? String
            val message: V2NIMChatroomMessage? = V2NIMChatroomMessageCreator.createTextMessage(text)
            if (message == null) {
                cont.resume(
                    NimResult(
                        code = createMessageFailed,
                        errorDetails = "create Message failed"
                    )
                )
            } else {
                cont.resume(NimResult(0, data = message.toMap()))
            }
        }
    }

    private suspend fun createImageMessage(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            val imagePath = arguments["imagePath"] as? String
            val name = arguments["name"] as? String
            val sceneName = arguments["sceneName"] as? String
            val height = arguments["height"] as? Int
            val width = arguments["width"] as? Int
            val message: V2NIMChatroomMessage? = V2NIMChatroomMessageCreator.createImageMessage(
                imagePath,
                name,
                sceneName,
                width,
                height
            )
            if (message == null) {
                cont.resume(
                    NimResult(
                        code = createMessageFailed,
                        errorDetails = "create Message failed"
                    )
                )
            } else {
                cont.resume(NimResult(0, data = message.toMap()))
            }
        }
    }

    private suspend fun createAudioMessage(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            val audioPath = arguments["audioPath"] as? String
            val name = arguments["name"] as? String
            val sceneName = arguments["sceneName"] as? String
            val duration = arguments["duration"] as? Int ?: 0
            val message: V2NIMChatroomMessage? = V2NIMChatroomMessageCreator.createAudioMessage(
                audioPath,
                name,
                sceneName,
                duration
            )
            if (message == null) {
                cont.resume(
                    NimResult(
                        code = createMessageFailed,
                        errorDetails = "create Message failed"
                    )
                )
            } else {
                cont.resume(NimResult(0, data = message.toMap()))
            }
        }
    }

    private suspend fun createVideoMessage(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            val videoPath = arguments["videoPath"] as? String
            val name = arguments["name"] as? String
            val sceneName = arguments["sceneName"] as? String
            val duration = arguments["duration"] as? Int ?: 0
            val height = arguments["height"] as? Int
            val width = arguments["width"] as? Int
            val message: V2NIMChatroomMessage? = V2NIMChatroomMessageCreator.createVideoMessage(
                videoPath,
                name,
                sceneName,
                duration,
                width,
                height
            )
            if (message == null) {
                cont.resume(
                    NimResult(
                        code = createMessageFailed,
                        errorDetails = "create Message failed"
                    )
                )
            } else {
                cont.resume(NimResult(0, data = message.toMap()))
            }
        }
    }

    private suspend fun createFileMessage(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            val filePath = arguments["filePath"] as? String
            val name = arguments["name"] as? String
            val sceneName = arguments["sceneName"] as? String
            val message: V2NIMChatroomMessage? = V2NIMChatroomMessageCreator.createFileMessage(filePath, name, sceneName)
            if (message == null) {
                cont.resume(
                    NimResult(
                        code = createMessageFailed,
                        errorDetails = "create Message failed"
                    )
                )
            } else {
                cont.resume(NimResult(0, data = message.toMap()))
            }
        }
    }

    private suspend fun createLocationMessage(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            val latitude = arguments["latitude"] as? Double ?: 0.0
            val longitude = arguments["longitude"] as? Double ?: 0.0
            val address = arguments["address"] as? String
            val message: V2NIMChatroomMessage? = V2NIMChatroomMessageCreator.createLocationMessage(latitude, longitude, address)
            if (message == null) {
                cont.resume(
                    NimResult(
                        code = createMessageFailed,
                        errorDetails = "create Message failed"
                    )
                )
            } else {
                cont.resume(NimResult(0, data = message.toMap()))
            }
        }
    }

    private suspend fun createCustomMessage(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            val rawAttachment = arguments["rawAttachment"] as? String
            val message: V2NIMChatroomMessage? = V2NIMChatroomMessageCreator.createCustomMessage(rawAttachment)
            if (message == null) {
                cont.resume(
                    NimResult(
                        code = createMessageFailed,
                        errorDetails = "create Message failed"
                    )
                )
            } else {
                cont.resume(NimResult(0, data = message.toMap()))
            }
        }
    }

    private suspend fun createCustomMessageWithAttachment(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            val rawAttachment = arguments["rawAttachment"] as? String
            val attachment = V2NIMMessageAttachmentCreator.createCustomMessageAttachment(rawAttachment) as? V2NIMMessageCustomAttachment
            val message: V2NIMChatroomMessage? = V2NIMChatroomMessageCreator.createCustomMessageWithAttachment(attachment)
            if (message == null) {
                cont.resume(
                    NimResult(
                        code = createMessageFailed,
                        errorDetails = "create Message failed"
                    )
                )
            } else {
                cont.resume(NimResult(0, data = message.toMap()))
            }
        }
    }

    private suspend fun createCustomMessageWithAttachmentAndSubType(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            val rawAttachment = arguments["attachment"] as? String
            val subType = arguments["subType"] as? Int
            if (subType == null) {
                cont.resume(
                    NimResult(
                        code = createMessageFailed,
                        errorDetails = "create CustomMessageWithAttachmentAndType subtype is null"
                    )
                )
            } else {
                val attachment = V2NIMMessageAttachmentCreator.createCustomMessageAttachment(rawAttachment) as? V2NIMMessageCustomAttachment
                val message: V2NIMChatroomMessage? = V2NIMChatroomMessageCreator.createCustomMessageWithAttachment(attachment, subType)
                if (message == null) {
                    cont.resume(
                        NimResult(
                            code = createMessageFailed,
                            errorDetails = "create Message failed"
                        )
                    )
                } else {
                    cont.resume(NimResult(0, data = message.toMap()))
                }
            }
        }
    }

    private suspend fun createTipsMessage(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            val text = arguments["text"] as? String
            val message: V2NIMChatroomMessage? = V2NIMChatroomMessageCreator.createTipsMessage(text)
            if (message == null) {
                cont.resume(
                    NimResult(
                        code = createMessageFailed,
                        errorDetails = "create Message failed"
                    )
                )
            } else {
                cont.resume(NimResult(0, data = message.toMap()))
            }
        }
    }

    private suspend fun createForwardMessage(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            val messageMap = arguments["message"] as Map<String, *>?
            val message = messageMap?.toV2NIMChatroomMessage()
            val messageResult: V2NIMChatroomMessage? = V2NIMChatroomMessageCreator.createForwardMessage(message)
            if (messageResult == null) {
                cont.resume(NimResult(0, data = mapOf()))
            } else {
                cont.resume(NimResult(0, data = messageResult.toMap()))
            }
        }
    }
}

File: nim_core_v2/ios/Classes/services/FLTChatRoomMessageCreatorService.swift
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

import NIMSDK

enum ChatRoomMessageCreatorAPIType: String {
  case createTextMessage
  case createImageMessage
  case createAudioMessage
  case createVideoMessage
  case createFileMessage
  case createLocationMessage
  case createCustomMessage
  case createCustomMessageWithAttachmentAndSubType
  case createForwardMessage
  case createTipsMessage
}

class FLTChatRoomMessageCreatorService: FLTBaseService, FLTService {
  func serviceName() -> String {
    ServiceType.ChatRoomMessageCreatorService.rawValue
  }

  func onMethodCalled(_ method: String, _ arguments: [String: Any], _ resultCallback: ResultCallback) {
    switch ChatRoomMessageCreatorAPIType(rawValue: method) {
    case .createTextMessage:
      createTextMessage(arguments, resultCallback)
    case .createImageMessage:
      createImageMessage(arguments, resultCallback)
    case .createAudioMessage:
      createAudioMessage(arguments, resultCallback)
    case .createVideoMessage:
      createVideoMessage(arguments, resultCallback)
    case .createFileMessage:
      createFileMessage(arguments, resultCallback)
    case .createLocationMessage:
      createLocationMessage(arguments, resultCallback)
    case .createCustomMessage:
      createCustomMessage(arguments, resultCallback)
    case .createCustomMessageWithAttachmentAndSubType:
      createCustomMessageWithAttachmentAndSubType(arguments, resultCallback)
    case .createForwardMessage:
      createForwardMessage(arguments, resultCallback)
    case .createTipsMessage:
      createTipsMessage(arguments, resultCallback)
    default:
      resultCallback.notImplemented()
    }
  }

  func register(_ nimCore: NimCore) {
    self.nimCore = nimCore
    nimCore.addService(self)
  }

  /// 创建文本消息
  func createTextMessage(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let text = arguments["text"] as? String else {
      parameterError(resultCallback)
      return
    }
    let message = V2NIMChatroomMessageCreator.createTextMessage(text)
    successCallBack(resultCallback, message.toDic())
  }

  /// 创建音频消息
  func createAudioMessage(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let path = arguments["audioPath"] as? String,
          let duration = arguments["duration"] as? Int32 else {
      parameterError(resultCallback)
      return
    }

    let name = arguments["name"] as? String
    var sceneName = arguments["sceneName"] as? String
    sceneName = sceneName ?? V2NIMStorageSceneConfig.default_IM().sceneName
    let message = V2NIMChatroomMessageCreator.createAudioMessage(path, name: name, sceneName: sceneName, duration: duration)
    successCallBack(resultCallback, message.toDic())
  }

  /// 创建图片消息
  func createImageMessage(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let imagePath = arguments["imagePath"] as? String,
          let width = arguments["width"] as? Int32,
          let height = arguments["height"] as? Int32 else {
      parameterError(resultCallback)
      return
    }
    let name = arguments["name"] as? String
    var sceneName = arguments["sceneName"] as? String
    sceneName = sceneName ?? V2NIMStorageSceneConfig.default_IM().sceneName
    let message = V2NIMChatroomMessageCreator.createImageMessage(imagePath, name: name, sceneName: sceneName, width: width, height: height)
    successCallBack(resultCallback, message.toDic())
  }

  /// 创建视频消息
  func createVideoMessage(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let videoPath = arguments["videoPath"] as? String,
          let duration = arguments["duration"] as? Int32,
          let width = arguments["width"] as? Int32,
          let height = arguments["height"] as? Int32 else {
      parameterError(resultCallback)
      return
    }
    let name = arguments["name"] as? String
    var sceneName = arguments["sceneName"] as? String
    sceneName = sceneName ?? V2NIMStorageSceneConfig.default_IM().sceneName
    let message = V2NIMChatroomMessageCreator.createVideoMessage(videoPath, name: name, sceneName: sceneName, duration: duration, width: width, height: height)
    successCallBack(resultCallback, message.toDic())
  }

  /// 创建文件消息
  func createFileMessage(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let filePath = arguments["filePath"] as? String else {
      parameterError(resultCallback)
      return
    }
    let name = arguments["name"] as? String
    var sceneName = arguments["sceneName"] as? String
    sceneName = sceneName ?? V2NIMStorageSceneConfig.default_IM().sceneName
    let message = V2NIMChatroomMessageCreator.createFileMessage(filePath, name: name, sceneName: sceneName)
    successCallBack(resultCallback, message.toDic())
  }

  /// 创建地理位置消息
  func createLocationMessage(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let latitude = arguments["latitude"] as? Double,
          let longitude = arguments["longitude"] as? Double,
          let address = arguments["address"] as? String else {
      parameterError(resultCallback)
      return
    }

    let message = V2NIMChatroomMessageCreator.createLocationMessage(latitude, longitude: longitude, address: address)
    successCallBack(resultCallback, message.toDic())
  }

  /// 创建自定义消息
  func createCustomMessage(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let rawAttachment = arguments["rawAttachment"] as? String else {
      parameterError(resultCallback)
      return
    }
    let message = V2NIMChatroomMessageCreator.createCustomMessage(rawAttachment)
    successCallBack(resultCallback, message.toDic())
  }

  /// 创建自定义消息
  func createCustomMessageWithAttachmentAndSubType(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let subType = arguments["subType"] as? Int32,
          let attachment = arguments["attachment"] as? String else {
      parameterError(resultCallback)
      return
    }

    let rawAttachment = V2NIMMessageCustomAttachment()
    rawAttachment.raw = attachment
    let message = V2NIMChatroomMessageCreator.createCustomMessage(with: rawAttachment, subType: subType)
    successCallBack(resultCallback, message.toDic())
  }

  /// 创建tips消息
  func createTipsMessage(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let text = arguments["text"] as? String else {
      parameterError(resultCallback)
      return
    }
    let message = V2NIMChatroomMessageCreator.createTipsMessage(text)
    successCallBack(resultCallback, message.toDic())
  }

  /// 创建转发消息
  func createForwardMessage(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let messageDic = arguments["message"] as? [String: Any] else {
      parameterError(resultCallback)
      return
    }
    let message = V2NIMChatroomMessage.fromDic(messageDic)
    let forwardMessage = V2NIMChatroomMessageCreator.createForwardMessage(message)
    successCallBack(resultCallback, forwardMessage.toDic())
  }
}
